feat: add new cli params

juliangieseke opened the pull request:
This adds new cli params for the data node (for now, will add more).

Im unsure if it should be handled like so (merging values into the default config) of if there should be some Config.new(args) constructor taking params as arguments? 

👀 appreciated 🙇

nfnt committed:
feat: add new cli params

Co-authored-by: Julian Gieseke <julian@katulu.io>

nfnt commented:
I refactored a few things in an additional commit. It's still building on the `Default` implementation of a `Config` but adds the overrides on top.

orlandohohmeier reviewed (commented):
I'm back and forth on whether it's ok to make all the config field public - one one hand it make things like this much easier but then also we loose some guarantees. Why can't we use the provider to augment the config with a tls config created based on a name?

nfnt commented:
IMO, these structures are simple data structures (aka [plain old data](https://en.wikipedia.org/wiki/Passive_data_structure)). There's no need for information hiding as there's no functionality specific to these structures.

juliangieseke commented:
needs rebase, lmk if I should go ahead rebase or if there is more discussion necessary (or feel free to just rebase)

orlandohohmeier commented:
> IMO, these structures are simple data structures (aka plain old data). There's no need for information hiding as there's no functionality specific to these structures.

It's not about simple data structures or not but that at the moment the Config is built using providers which include information about the source of an option i.e which provider they came from. For the init case it absolutely doesn't matter that they're pub but for their other cases we loose the guarantee that we can tell where an option came from. I think we could achieve the same using providers but I'm not block this one - it just means that we need to be super cautious when reviewing config related code to make sure that folks use providers in the non init case.

nfnt commented:
I'll look into using a `figment::Provider` for the command line arguments, as it seems the easiest way to merge the overridden values. Right now, one has to provide a complete `Job` as JSON, but we want users to only provide the parts of a `Job` that need to be overwritten.

orlandohohmeier commented:
Just to be super clear, I didn't block this one and I'd okay to merge this as is considering that this is currently blocking @juliangieseke and then _fix_ this later if we see need to do so.

juliangieseke commented:
alright, rebased and ready for review, PTAL.

nfnt committed:
feat: change job defaults

Changes the default of a job to use our HuggingFace repository and
configuration for the MNIST demo.

nfnt committed:
feat: merge overrides with defaults

Uses 'figment' to merge command line argument overrides in the 'init'
subcommand with our default configuration. Notably this allows to
provide a subset of a 'SchedulerConfig' instead of a complete one when
using the '--job' argument.

nfnt committed:
Merge pull request #127 from hypha-space/nfnt/finer-defaults

Merge command line arguments with defaults

juliangieseke commented on crates/data/src/bin/hypha-data.rs:
```diff
@@ -226,8 +229,31 @@ async fn run(config: ConfigWithMetadata<Config>) -> Result<()> {
+
+            let config = config_builder.build()?.validate()?;
+
+            fs::write(output, &to_toml(&config.config).into_diagnostic()?)
```
seems like you droppe having the name in the filename. (true for all clis)

juliangieseke replied:
that was actually a bug... `-n` woudl have overwritten `-o`
